packages/react-components/src/components/xy-output-component.tsx

```diff
@@ -809,8 +844,22 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
                     break;
                 }
                 case 'Shift': {
-                    if (!this.isRightClick) {
-                        this.setState({ cursor: 'crosshair', shiftKey: true });
+                    if (!this.isPanning && !this.isRightClick && !this.isSelecting) {
+                        if (key.ctrlKey) {
+                            this.setState({ cursor: 'default' });
+                        } else {
+                            this.setState({ cursor: 'crosshair', shiftKey: true });
+                        }
+                    }
+                    break;
+                }
+                case 'Control': {
+                    if (!this.isSelecting && !this.isPanning) {
+                        if (key.shiftKey) {
+                            this.setState({ cursor: 'default' });
+                        } else if (!this.state.shiftKey) {
+                            this.setState({ cursor: 'grabbing', ctrlKey: true });
+                        }
```

Comment:
The shiftKey and ctrlKey state is not updated properly because it depends on other conditions than just the key being pressed.

Reply:
Actually, I don't think these are needed to be added to the state because these modifier key states are part of every KeyEvent or MouseEvent? Better use that directly than rely on our own copy...

Reply:
Use case: Hold mouse down, hold Shift down, release mouse, cursor should remain crosshair.
Use case: Hold Ctrl down, hold mouse down, release Ctrl, hold Shift, release mouse, cursor should change to crosshair.
Use case: Hold Ctrl down, hold mouse down, hold Ctrl+Shift down, release mouse, cursor should change to default.

Reply:
Fixed.